Tools/machines/desktop/spack-ubuntu-cuda.yaml

```diff
@@ -15,11 +15,12 @@
 spack:
   specs:
   - adios2
+  - ascent +adios2 +python ~fortran
```

Comment:
@cyrush recommends to add static builds for DevilRay and VTK-m.
Maybe that also helps with
```
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIZN25PhysicalParticleContainer9AddPlasmaEiNS_7RealBoxEEUliiiRKNS_12RandomEngineEE_EENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT_vEE5valueEvE4typeERKNS_3BoxEOS9_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
```
?

Reply:
@cyrush @kwryankrattiger with the static lib requirement, I wonder if one of the packages in the vis stack (Ascent, VTK-m, ...) forgets to forward CUDA RDC flags via CMake or so?

Reply:
Ascent handles all of this (it was a major focus a few years ago). We also had all of the spack pacages in balance, and this is how we test compiling and linking CUDA in CI. 

Doesn't mean things don't break occasionally :-)

With VTK-m 1.8 shared is now possible, but that won't be supported until Ascent 0.9.0.

Reply:
The SDK prefers shared libs because it is less weird to link.

Reply:
nah, problem is not resolved with `~static` as currently pushed - @cyrush any more ideas? :)

Reply:
Iterations at my P100 build are going painfully slow.  

I am trying an experiment with sm=60 (tesla) instead of sm=61 (consumer) I hope to see it finish today.

I think what is happening here is that the composition of ascent stuff and warpx cude stuff when linked together are spilling over a device limit for pascal consumer. 

We are setting this flag in Ascent for our use of VTK-m:

(ascent develop)
https://github.com/Alpine-DAV/ascent/blob/deef65e39f3b2792a40281439c4f614488349c0b/src/cmake/thirdparty/SetupVTKm.cmake#L45

(vtk-h alone, which supports ascent 0.8.0)

Since we are compiling the vtk-m filters, that should take care of in Ascent. But again, maybe we are just hitting some limit when everything is pulled together.

Reply:
If I compile everything for SM 7.0, then things link.

Reply:
officially

```
VTKm CUDA Architecture Specifies what GPU architecture(s) to build CUDA for. The options include
native, fermi, kepler, maxwell, pascal, volta, and turing
```

Reply:
Thank you :) Ui, maybe a bit unlikely Fermi & Kepler are still working, not even the CUDA Toolkit supports those anymore :D

Reply:
Ahh - that's correct VTK-m will compile for all pascal variants. 

But when WarpX goes to link - I expect it will be providing a single sm case as part of its linking args.  Ascent may as well.  Many moving parts here - I am not sure what survives and what is gone as we travel down the rapids. 

My hypothesis is that the composed software might be too beefy for consumer pascal w/o `-Xptxas --disable-optimizer-constants`. We hit a hardware resource wall.  

I wish I had a good answer to verify specifically (cuobject dump like inspection, etc).  But even if we do have the exact answer of why, the prescription will still be to use `-Xptxas --disable-optimizer-constants`.

My pascal (sm 60 model) builds are painfully slow, which is adding to delay sorry :-(

Reply:
I was able to link when (and only when) I used `-Xptxas --disable-optimizer-constants.`

However I can't run:

```
STEP 1 starts ...
amrex::Abort::0::GPU last error detected in file /usr/WS1/cyrush/2022.08.warpx/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_GpuLaunchFunctsG.H line 1053: invalid device function !!!
SIGABRT
```
```
cat Backtrace.0 
=== If no file names and line numbers are shown below, one can run
            addr2line -Cpfie my_exefile my_line_address
    to convert `my_line_address` (e.g., 0x4a6b) into file name and line number.
    Or one can use amrex/Tools/Backtrace/parse_bt.py.

=== Please note that the line number reported by addr2line may not be accurate.
    One can use
            readelf -wl my_exefile | grep my_line_address'
    to find out the offset for that line.

 0: ./warpx() [0x8d80de]
    amrex::BLBackTrace::print_backtrace_info(_IO_FILE*) at ??:?

 1: ./warpx() [0x8d96cb]
    amrex::BLBackTrace::handler(int) at ??:?

 2: ./warpx() [0x4b5ce9]
    amrex::Gpu::ErrorCheck(char const*, int) at ??:?

 3: ./warpx() [0x6c881c]
    void FiniteDifferenceSolver::EvolveBCartesian<CartesianYeeAlgorithm>(std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul>&, std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul> const&, std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> > const&, int, double) at ??:?

 4: ./warpx() [0x6c73da]
    FiniteDifferenceSolver::EvolveB(std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul>&, std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul> const&, std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> > const&, std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul> const&, std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul> const&, std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul>&, std::array<std::unique_ptr<amrex::MultiFab, std::default_delete<amrex::MultiFab> >, 3ul>&, std::array<std::unique_ptr<amrex::iMultiFab, std::default_delete<amrex::iMultiFab> >, 3ul>&, std::array<std::unique_ptr<amrex::LayoutData<FaceInfoBox>, std::default_delete<amrex::LayoutData<FaceInfoBox> > >, 3ul>&, int, double) at ??:?

 5: ./warpx() [0x6b71cf]
    WarpX::EvolveB(int, PatchType, double, DtType) at ??:?

 6: ./warpx() [0x6b7372]
    WarpX::EvolveB(int, double, DtType) at ??:?

 7: ./warpx() [0x6b7406]
    WarpX::EvolveB(double, DtType) at ??:?

 8: ./warpx() [0x69fece]
    WarpX::OneStep_nosub(double) at ??:?

 9: ./warpx() [0x6a26b8]
    WarpX::Evolve(int) at ??:?

10: ./warpx() [0x48921d]
    main at ??:?

11: /lib64/libc.so.6(__libc_start_main+0xf5) [0x2aaabf588555]
    __libc_start_main
../csu/libc-start.c:300

12: ./warpx() [0x49979e]
    _start at ??:?
```

I think this is before we get to Ascent.

Reply:
But it could still be sad about something in Ascent.

Reply:
Yep, maybe we cannot handle that flag well... `invalid device function`, curious. could be the last kernel called before